File: Paginas/servicios.py
import flet as ft
from Controladores_bases.Controlador import Servicio, EstadoServicio

db = 'cjs.db'
servicio_controller = Servicio(db)
estado_controller = EstadoServicio(db)

# Página de administración de servicios
def administracion_servicios(page):
    def mostrar_historial_servicios():
        try:
            servicios = servicio_controller.obtener_todos_los_servicios()
            return [
                ft.Card(
                    content=ft.Container(
                        content=ft.Column([
                            ft.ListTile(
                                leading=ft.Icon(ft.icons.WORK),
                                title=ft.Text(servicio['nombre'], weight=ft.FontWeight.BOLD),
                                subtitle=ft.Text(f"Fecha: {servicio['fecha']} | Detalles: {servicio['detalles']}"),
                            ),
                            ft.Row([
                                ft.ElevatedButton("Editar", icon=ft.icons.EDIT,
                                                  on_click=lambda _, id=servicio['id_servicio']: editar_servicio(id)),
                                ft.ElevatedButton("Eliminar", icon=ft.icons.DELETE,
                                                  on_click=lambda _, id=servicio['id_servicio']: eliminar_servicio(id)),
                            ], alignment=ft.MainAxisAlignment.END)
                        ]),
                        padding=10
                    ),
                    margin=10
                ) for servicio in servicios
            ]
        except Exception as e:
            print(f"Error al obtener servicios: {e}")
            return []

    lista_servicios = ft.Column([], scroll=ft.ScrollMode.AUTO, expand=True)

    def actualizar_lista():
        lista_servicios.controls = mostrar_historial_servicios()
        page.update()

    # Inputs para crear/editar servicios
    nombre_input = ft.TextField(label="Nombre del Servicio", expand=True)
    fecha_input = ft.TextField(label="Fecha (YYYY-MM-DD)", expand=True)
    telefono_input = ft.TextField(label="Teléfono Extra", expand=True)
    detalles_input = ft.TextField(label="Detalles del Servicio", expand=True)
    estado_id_input = ft.TextField(label="ID del Estado", expand=True)
    #cliente_id_input = ft.TextField(label="ID del Cliente", expand=True)
    #personal_id_input = ft.TextField(label="ID del Personal", expand=True)
    message = ft.Text()

    def agregar_servicio(e):
        nombre = nombre_input.value
        fecha = fecha_input.value
        telefono_extra = telefono_input.value
        detalles = detalles_input.value
        estado_id = estado_id_input.value
        cliente_id = '1'
        personal_id = '1'

        if servicio_controller.crear_servicio(nombre, fecha, telefono_extra, detalles, estado_id, cliente_id, personal_id):
            message.value = "Servicio creado exitosamente."
            nombre_input.value = ""
            fecha_input.value = ""
            telefono_input.value = ""
            detalles_input.value = ""
            estado_id_input.value = ""
            #cliente_id_input.value = ""
            #personal_id_input.value = ""
            actualizar_lista()
        else:
            message.value = "Error al crear el servicio."
        page.update()

    def editar_servicio(id_servicio):
        servicio = servicio_controller.obtener_servicio(id_servicio)
        if servicio:
            nombre_input.value = servicio['nombre']
            fecha_input.value = servicio['fecha']
            telefono_input.value = servicio['telefono_extra']
            detalles_input.value = servicio['detalles']
            estado_id_input.value = servicio['estado_id']
            #cliente_id_input.value = servicio['cliente_id']
            #personal_id_input.value = servicio['personal_id']
        page.update()

    def eliminar_servicio(id_servicio):
        servicio_controller.eliminar_servicio(id_servicio)
        message.value = "Servicio eliminado."
        actualizar_lista()

    vista = ft.View(
        "/administrar_servicios",
        [
            ft.AppBar(title=ft.Text("Administración de Servicios"), bgcolor=ft.colors.BLUE),
            ft.Container(
                content=ft.Column([
                    ft.Text("Agregar Nuevo Servicio", size=20, weight=ft.FontWeight.BOLD),
                    ft.Row([nombre_input, fecha_input, telefono_input], expand=True),
                    ft.Row([detalles_input, estado_id_input], expand=True),
                    ft.ElevatedButton("Agregar Servicio", on_click=agregar_servicio, icon=ft.icons.ADD),
                    message,
                ]),
                padding=20,
            ),
            ft.Divider(),
            ft.Text("Servicios Registrados", size=20, weight=ft.FontWeight.BOLD),
            lista_servicios,
        ],
        scroll=ft.ScrollMode.AUTO
    )

    actualizar_lista()
    return vista

File: Flutter/Cjsapp/lib/Paginas/servicios.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:Cjsapp/Modelos/principal.dart';
import 'package:Cjsapp/Controladores/base_principal_controlador.dart';

class AdministracionServicios extends StatefulWidget {
  const AdministracionServicios({super.key});

  @override
  _AdministracionServiciosState createState() => _AdministracionServiciosState();
}

class _AdministracionServiciosState extends State<AdministracionServicios> {
  final _formKey = GlobalKey<FormState>();
  bool modoEdicion = false;
  Servicio? servicioActual;
  
  final nombreController = TextEditingController();
  final fechaController = TextEditingController();
  final telefonoController = TextEditingController();
  final detallesController = TextEditingController();
  int? estadoSeleccionado;
  int? clienteSeleccionado;
  int? empleadoSeleccionado;
  
  List<EstadoServicio> estados = [];
  List<Cliente> clientes = [];
  List<Empleado> empleados = [];

  @override
  void initState() {
    super.initState();
    cargarDatos();
    fechaController.text = DateFormat('yyyy-MM-dd').format(DateTime.now());
  }

  void cargarDatos() async {
    estados = await DatabaseController.obtenerTodosLosEstados();
    clientes = await DatabaseController.obtenerTodosLosClientes();
    empleados = await DatabaseController.obtenerTodosLosEmpleados();
    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Administración de Servicios"),
        backgroundColor: const Color(0xFF316938),
        actions: [
          IconButton(
            icon: const Icon(Icons.home),
            onPressed: () => Navigator.pushNamed(context, '/home'),
          ),
          IconButton(
            icon: const Icon(Icons.settings),
            onPressed: () => Navigator.pushNamed(context, '/estado_servicios'),
          ),
        ],
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            Padding(
              padding: const EdgeInsets.all(16.0),
              child: Form(
                key: _formKey,
                child: Column(
                  children: [
                    TextFormField(
                      controller: nombreController,
                      decoration: const InputDecoration(labelText: "Nombre del Servicio"),
                      validator: (value) {
                        if (value == null || value.isEmpty) {
                          return 'Por favor ingrese un nombre';
                        }
                        return null;
                      },
                    ),
                    TextFormField(
                      controller: fechaController,
                      decoration: const InputDecoration(labelText: "Fecha (YYYY-MM-DD)"),
                      validator: (value) {
                        if (value == null || value.isEmpty) {
                          return 'Por favor ingrese una fecha';
                        }
                        // Add date format validation if needed
                        return null;
                      },
                    ),
                    TextFormField(
                      controller: telefonoController,
                      decoration: const InputDecoration(labelText: "Teléfono Extra"),
                    ),
                    TextFormField(
                      controller: detallesController,
                      decoration: const InputDecoration(labelText: "Detalles del Servicio"),
                    ),
                    DropdownButtonFormField<int>(
                      value: estadoSeleccionado,
                      items: estados.map((estado) {
                        return DropdownMenuItem(
                          value: estado.idEstado,
                          child: Text(estado.descripcion),
                        );
                      }).toList(),
                      onChanged: (value) {
                        setState(() {
                          estadoSeleccionado = value;
                        });
                      },
                      decoration: const InputDecoration(labelText: "Estado del Servicio"),
                      validator: (value) {
                        if (value == null) {
                          return 'Por favor seleccione un estado';
                        }
                        return null;
                      },
                    ),
                    DropdownButtonFormField<int>(
                      value: clienteSeleccionado,
                      items: clientes.map((cliente) {
                        return DropdownMenuItem(
                          value: cliente.idCliente,
                          child: Text(cliente.nombre),
                        );
                      }).toList(),
                      onChanged: (value) {
                        setState(() {
                          clienteSeleccionado = value;
                        });
                      },
                      decoration: const InputDecoration(labelText: "Cliente"),
                      validator: (value) {
                        if (value == null) {
                          return 'Por favor seleccione un cliente';
                        }
                        return null;
                      },
                    ),
                    DropdownButtonFormField<int>(
                      value: empleadoSeleccionado,
                      items: empleados.map((empleado) {
                        return DropdownMenuItem(
                          value: empleado.idPersonal,
                          child: Text(empleado.nombre),
                        );
                      }).toList(),
                      onChanged: (value) {
                        setState(() {
                          empleadoSeleccionado = value;
                        });
                      },
                      decoration: const InputDecoration(labelText: "Empleado"),
                      validator: (value) {
                        if (value == null) {
                          return 'Por favor seleccione un empleado';
                        }
                        return null;
                      },
                    ),
                    const SizedBox(height: 20),
                    ElevatedButton(
                      onPressed: guardarOActualizarServicio,
                      child: Text(modoEdicion ? "Actualizar Servicio" : "Agregar Servicio"),
                    ),
                  ],
                ),
              ),
            ),
            const Divider(),
            const Text("Servicios Registrados", style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold)),
            FutureBuilder<List<Servicio>>(
              future: DatabaseController.obtenerTodosLosServicios(),
              builder: (context, snapshot) {
                if (snapshot.connectionState == ConnectionState.waiting) {
                  return const Center(child: CircularProgressIndicator());
                } else if (snapshot.hasError) {
                  return Center(child: Text("Error: ${snapshot.error}"));
                } else if (!snapshot.hasData || snapshot.data!.isEmpty) {
                  return const Center(child: Text("No hay servicios registrados"));
                } else {
                  return ListView.builder(
                    shrinkWrap: true,
                    physics: const NeverScrollableScrollPhysics(),
                    itemCount: snapshot.data!.length,
                    itemBuilder: (context, index) {
                      final servicio = snapshot.data![index];
                      return Card(
                        margin: const EdgeInsets.all(8),
                        child: ListTile(
                          title: Text(servicio.nombre),
                          subtitle: Text("Fecha: ${DateFormat('yyyy-MM-dd').format(servicio.fecha)} | Detalles: ${servicio.detalles ?? ''}"),
                          trailing: Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              IconButton(
                                icon: const Icon(Icons.edit),
                                onPressed: () => editarServicio(servicio),
                              ),
                              IconButton(
                                icon: const Icon(Icons.delete),
                                onPressed: () => eliminarServicio(servicio.idServicio!),
                              ),
                            ],
                          ),
                        ),
                      );
                    },
                  );
                }
              },
            ),
          ],
        ),
      ),
    );
  }

  void guardarOActualizarServicio() {
    if (_formKey.currentState!.validate()) {
      final nombre = nombreController.text;
      final fecha = DateTime.parse(fechaController.text);
      final telefonoExtra = telefonoController.text;
      final detalles = detallesController.text;

      if (modoEdicion && servicioActual != null) {
        final servicioActualizado = servicioActual!.copyWith(
          nombre: nombre,
          fecha: fecha,
          telefonoExtra: telefonoExtra,
          detalles: detalles,
          estadoId: estadoSeleccionado!,
          clienteId: clienteSeleccionado!,
          personalId: empleadoSeleccionado!,
        );
        DatabaseController.actualizarServicio(servicioActualizado);
      } else {
        final nuevoServicio = Servicio(
          nombre: nombre,
          fecha: fecha,
          telefonoExtra: telefonoExtra,
          detalles: detalles,
          estadoId: estadoSeleccionado!,
          clienteId: clienteSeleccionado!,
          personalId: empleadoSeleccionado!,
        );
        DatabaseController.crearServicio(nuevoServicio);
      }

      limpiarFormulario();
      setState(() {
        modoEdicion = false;
        servicioActual = null;
      });
    }
  }

  void editarServicio(Servicio servicio) {
    setState(() {
      modoEdicion = true;
      servicioActual = servicio;
      nombreController.text = servicio.nombre;
      fechaController.text = DateFormat('yyyy-MM-dd').format(servicio.fecha);
      telefonoController.text = servicio.telefonoExtra ?? '';
      detallesController.text = servicio.detalles ?? '';
      estadoSeleccionado = servicio.estadoId;
      clienteSeleccionado = servicio.clienteId;
      empleadoSeleccionado = servicio.personalId;
    });
  }

  void eliminarServicio(int idServicio) async {
    await DatabaseController.eliminarServicio(idServicio);
    setState(() {});
  }

  void limpiarFormulario() {
    nombreController.clear();
    fechaController.text = DateFormat('yyyy-MM-dd').format(DateTime.now());
    telefonoController.clear();
    detallesController.clear();
    estadoSeleccionado = null;
    clienteSeleccionado = null;
    empleadoSeleccionado = null;
  }

  @override
  void dispose() {
    nombreController.dispose();
    fechaController.dispose();
    telefonoController.dispose();
    detallesController.dispose();
    super.dispose();
  }
}